tensorflow_recommenders_addons/dynamic_embedding/core/kernels/redis_impl/redis_cluster_connection_pool.hpp

```diff
@@ -0,0 +1,757 @@
+    pool_opts.size = redis_connection_params.redis_conn_pool_size;
+    pool_opts.wait_timeout =
+        std::chrono::milliseconds(redis_connection_params.redis_wait_timeout);
+    pool_opts.connection_lifetime =
+        std::chrono::minutes(redis_connection_params.redis_connection_lifetime);
+
+    try {
+      static auto redis_client =
+          std::make_shared<RedisInstance>(RedisInstance(conn_opts, pool_opts));
+      redis_client->set("key test for connecting", "val test for connecting",
+                        std::chrono::milliseconds(1));
+      return redis_client;
+    } catch (const std::exception &err) {
+      std::cerr << "RedisHandler--other " << err.what() << std::endl;
+      return nullptr;
+    } catch (...) {
+      std::cerr << "RedisHandler--other crash" << std::endl;
+      return nullptr;
+    }
+    return nullptr;
+  }
+
+  virtual void conn() override {
+    if (isRedisConnect == false) {
+      for (short i = 0; i < 10; i++) {
+        redis_conn = start_conn();
+        if (redis_conn) {
+          isRedisConnect = true;
+          break;
+        }
+      }
+      if (isRedisConnect == false) {
+        std::cerr << "Can not connect to the Redis Cluster servers."
+                  << std::endl;
+        throw(std::runtime_error("Exit without any Redis connection."));
+      }
+    }
+  }
+
+  static std::shared_ptr<RedisWrapper<RedisInstance, K, V>> get_instance() {
+    // for the Meyer's Singleton mode
+    static std::shared_ptr<RedisWrapper<RedisInstance, K, V>> instance_ptr =
+        std::make_shared<RedisWrapper<RedisInstance, K, V>>();
+    return instance_ptr;
+  }
+
+private:
+  template <typename Cmd>
+  std::vector<std::unique_ptr<redisReply, ::sw::redis::ReplyDeleter>>
+  pipe_exec(Cmd cmd, const unsigned &storage_slice, const unsigned &size_check,
+            const ThreadContext &thread_context) {
+    std::vector<std::unique_ptr<redisReply, ::sw::redis::ReplyDeleter>> replies;
+    replies.reserve(storage_slice);
+    for (unsigned i = 0; i < storage_slice; ++i) {
+      if (thread_context.slots[i].ptrs.size() >= size_check) {
+        ::sw::redis::StringView hkey(thread_context.slots[i].ptrs[1],
+                                     thread_context.slots[i].sizes[1]);
+        try {
+          replies.push_back(redis_conn->command(cmd, hkey,
+                                                thread_context.slots[i].ptrs,
+                                                thread_context.slots[i].sizes));
+        } catch (const std::exception &err) {
+          std::cerr << "RedisHandler error in pipe_exec for slices "
+                    << hkey.data() << " -- " << err.what() << std::endl;
+        }
+      } else {
+        replies.push_back(nullptr);
+      }
+    }
+    return replies;
+  }
+
+public:
+  virtual bool check_slices_num(const std::string &keys_prefix_name) override {
+    std::string redis_command = "KEYS " + keys_prefix_name + "[0123456789]";
```

Comment:
Why set [0123456789]?

Reply:
Already changed to [0123456789]*, There seems to be no better way.